lib/perl/Genome/Model/Tools/Bsmap/methylation_ratio.py

```diff
@@ -0,0 +1,181 @@
+        out_queue.put(counter)
+    print 'Quitting work'
+
+class Counter:
+    def __init__(self):
+        self.coverage = 0
+        self.methylation_coverage = 0
+        self.un_methylation_coverage = 0
+    def set_chromosome(self, chr):
+        self.chromosome = chr
+    def set_coverage(self, cov):
+        self.coverage = cov
+    def set_context(self, context):
+        self.context = context
+    def set_methylation_coverage(self, cov):
+        self.methylation_coverage = cov
+    def set_un_methylation_coverage(self, cov):
+        self.un_methylation_coverage = cov
+    def getCoverage(self):
+        return self.coverage
+    def getMethylationCoverage(self):
+        return self.methylation_coverage
+    def get_un_methylation_coverage(self):
+        return self.un_methylation_coverage
+    def get_chromosome(self):
+        return self.chromosome
+    def get_context(self):
+        return self.context
+
+if __name__ == '__main__':
+     
+    do_debug = False
+    usage = "usage: %prog [options] BSMAP_MAPPING_FILES (sorted BAM files)"
+    parser = optparse.OptionParser(usage=usage)
+    parser.add_option("-o", "--out", dest="outfile", metavar="FILE", help="output file name. (required)", default="")
+    parser.add_option("-d", "--ref", dest="reffile", metavar="FILE", help="reference genome fasta file. (required)", default="")
+    parser.add_option("-t", "--threads", dest="threads", metavar="THREADS", help="number of threads", default="1")
+    options, infile = parser.parse_args()
+    
+    if len(options.reffile) == 0: parser.error("Missing reference file, use -d or --ref option.")
+    if len(options.outfile) == 0: parser.error("Missing output file name, use -o or --out option.")
+    if len(infile) == 0: parser.error("Require at least one BSMAP_MAPPING_FILE.")
+    
+    
+    step = 100000
+    genome = pysam.FastaFile(options.reffile)
+    work_queue = Queue()
+    work_units = 0
+    for chromosome in genome.references:
+        chr_length = genome.get_reference_length(chromosome)
+        start = 0
+        end = min(start + step, chr_length)
+        while True:
+            work_queue.put([chromosome, start, end])
+            work_units += 1
+            if end == chr_length: break
+            start += step
+            end = min(start + step, chr_length)
+    
+    num_threads = int(options.threads)
+    write_queue = Queue(100)
+    for i in range(num_threads):
+        worker = Process(target=process_region_pileup, args=(i, infile[0], options.reffile, work_queue, write_queue,))
+        worker.daemon = True
+        worker.start()
+    
+    raw_file = tempfile.NamedTemporaryFile()
+    writer = Process(target=write_data, args=(write_queue, raw_file.name, work_units,))
+    writer.start()
+    writer.join()
+    print "Sorting and compressing reads..."
+    command = ''.join(['tail -n +2 ', raw_file.name, ' | sort -k1V -k2n | bgzip > ', options.outfile, '.bgz'])
```

Comment:
But the shell strips the quotes so when you pass it on to a new shell command here you would need to re-add them or rather there is probably an API that allows you to pass a list/array so you don't have to worry about escaping.  In Perl, this would be [IPC::Run](https://metacpan.org/pod/IPC::Run).

What I would propose here is either support spaces or check if `out_file` has spaces early in the program and error if it does.